probe-rs-cli-util/src/common_options.rs

```diff
@@ -0,0 +1,476 @@
+    #[structopt(name = "speed", long = "speed", help = "The protocol speed in kHz.")]
+    pub speed: Option<u32>,
+    #[structopt(long = "dry-run")]
+    pub dry_run: bool,
+}
+
+static mut PROBE: Option<Probe> = None;
+static mut TARGET_SELECTOR: Option<TargetSelector> = None;
+
+impl ProbeOptions {
+    /// Add targets contained in file given by --chip-description-path
+    /// to probe-rs registery.
+    pub fn maybe_load_chip_desc(&self) -> Result<(), OperationError> {
+        if let Some(ref cdp) = self.chip_description_path {
+            probe_rs::config::add_target_from_yaml(&Path::new(cdp)).map_err(|error| {
+                OperationError::FailedChipDescriptionParsing {
+                    source: error,
+                    path: cdp.clone(),
+                }
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn attach_to_probe(&self) -> Result<&Probe, OperationError> {
+        // Tries to open the debug probe from the given commandline
+        // arguments. This ensures that there is only one probe
+        // connected or if multiple probes are found, a single one is
+        // specified via the commandline parameters.
+        let mut probe = {
+            if self.dry_run {
+                Probe::from_specific_probe(Box::new(FakeProbe::new()));
+            }
+
+            // If we got a probe selector as an argument, open the probe
+            // matching the selector if possible.
+            match &self.probe_selector {
+                Some(selector) => {
+                    Probe::open(selector.clone()).map_err(OperationError::FailedToOpenProbe)
+                }
+                None => {
+                    // Only automatically select a probe if there is
+                    // only a single probe detected.
+                    let list = Probe::list_all();
+                    if list.len() > 1 {
+                        return Err(OperationError::MultipleProbesFound { number: list.len() });
+                    }
+
+                    if let Some(info) = list.first() {
+                        Probe::open(info).map_err(OperationError::FailedToOpenProbe)
+                    } else {
+                        Err(OperationError::NoProbesFound)
+                    }
+                }
+            }
+        }?;
+
+        // Select protocol and speed
+        probe.select_protocol(self.protocol).map_err(|error| {
+            OperationError::FailedToSelectProtocol {
+                source: error,
+                protocol: self.protocol,
+            }
+        })?;
+        if let Some(speed) = self.speed {
+            let _actual_speed = probe.set_speed(speed).map_err(|error| {
+                OperationError::FailedToSelectProtocolSpeed {
+                    source: error,
+                    speed,
+                }
+            })?;
+        }
+
+        unsafe {
+            PROBE = Some(probe);
+            Ok(PROBE.as_ref().unwrap())
+        }
```

Comment:
Unsure how unsafe this is. `cargo-flash` wants to record some metadata for error reporting purposes, but I don't think the lifetime of this reference it tracked.